File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/testhelper/ContextBuilder.java
package uk.nhs.nhsx.testhelper;

import com.amazonaws.services.lambda.runtime.ClientContext;
import com.amazonaws.services.lambda.runtime.CognitoIdentity;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;

import java.nio.charset.StandardCharsets;

public class ContextBuilder {
    public static ContextBuilder context() {
        return new ContextBuilder();
    }

    public static Context aContext() {
        return new ContextBuilder().build();
    }

    public Context build() {
        return new TestContext();
    }

    public static class TestContext implements Context {
        @Override
        public String getAwsRequestId() {
            throw new UnsupportedOperationException("james didn't write");
        }

        @Override
        public String getLogGroupName() {
            return "test-log-group";
        }

        @Override
        public String getLogStreamName() {
            return "test-log-stream";
        }

        @Override
        public String getFunctionName() {
            return "function-name";
        }

        @Override
        public String getFunctionVersion() {
            return "function-version";
        }

        @Override
        public String getInvokedFunctionArn() {
            return "lambda-function-arn";
        }

        @Override
        public CognitoIdentity getIdentity() {
            return null;
        }

        @Override
        public ClientContext getClientContext() {
            throw new UnsupportedOperationException("james didn't write");
        }

        @Override
        public int getRemainingTimeInMillis() {
            return 1000;
        }

        @Override
        public int getMemoryLimitInMB() {
            return 100;
        }

        @Override
        public LambdaLogger getLogger() {
            return new LambdaLogger() {
                @Override
                public void log(String s) {
                    System.out.println("s = " + s);
                }

                @Override
                public void log(byte[] bytes) {
                    System.out.println("bytes = " + new String(bytes, StandardCharsets.UTF_8));
                }
            };
        }
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/testhelper/ContextBuilder.kt
package uk.nhs.nhsx.testhelper

import com.amazonaws.services.lambda.runtime.ClientContext
import com.amazonaws.services.lambda.runtime.CognitoIdentity
import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.LambdaLogger
import java.nio.charset.StandardCharsets
import java.util.UUID

class ContextBuilder {

    fun build(): Context = TestContext()

    class TestContext : Context by proxy() {
        override fun getAwsRequestId() = UUID(0,0).toString()

        override fun getLogGroupName(): String = "test-log-group"

        override fun getLogStreamName(): String = "test-log-stream"

        override fun getFunctionName(): String = "function-name"

        override fun getFunctionVersion(): String = "function-version"

        override fun getInvokedFunctionArn(): String = "lambda-function-arn"

        override fun getIdentity(): CognitoIdentity? = null

        override fun getRemainingTimeInMillis(): Int = 1000

        override fun getMemoryLimitInMB(): Int = 100

        override fun getLogger() = object : LambdaLogger {
            override fun log(s: String) {
                println("s = $s")
            }

            override fun log(bytes: ByteArray) {
                println("bytes = " + String(bytes, StandardCharsets.UTF_8))
            }
        }
    }

    companion object {
        fun context(): ContextBuilder = ContextBuilder()

        fun aContext(): Context = ContextBuilder().build()
    }
}
